### ISSUE TITLE
COPY optimization thoughts for phantom drives

### ISSUE THREAD

**Issue by boeckmann** [SvarDOS/bugz#135]
While thinking about the phantom drive thing I thought of some copy improvements to reduce the number of disk swaps needed:

1. After opening the input file, read the first block of data BEFORE opening the output file.
2. Close the output file before the input file if the last IO operation was on the output file (write).

This should optimally save two swaps or even more.

**Comment by mateuszviste** [comment]
These ideas are on my radar already and they are the reason why I keep #123 open, but I am a bit reluctant to implement them for the time being as it would make the code much more spaghetti than it already is. COPY needs to handle a few weird cases like 'COPY A+*.TXT' ("append A to itself and then append all txt files to it") which makes the logic not as straightforward as I'd like it to be. COPY needs some refactoring first, possibly rewriting its inline asm to C. That's something I plan to look at after 2024.7 is out.

Another optimization is to read the source's date/time when opening it and keep the information in RAM, to avoid a swap at the end of the copy when datetime needs to be applied on dst before closing it.

Also using TRUENAME to validate/sanitize the src and dst filenames should be removed because truename on a phantom drive triggers a swap.

All these changes should lead to copy requiring only two swaps to copy a 60K file (same as MSDOS).

**Comment by boeckmann** [comment]
> All these changes should lead to copy requiring only two swaps to copy a 60K file (same as MSDOS).

Yes, two would be the lower bound and could certainly be achieved if the file fits into the buffer. So  closing this if its on your radar?

**Comment by mateuszviste** [comment]
I have closed #123 with the "easy wins" so let's keep this one open for memory.
Thanks for your thoughts. :)
